src/Form/LinkBlockFormDataProvider.php

```diff
@@ -0,0 +1,272 @@
+    }
+
+    /**
+     * @return array
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    public function getData()
+    {
+        if (null === $this->idLinkBlock) {
+            return [];
+        }
+
+        $linkBlock = new LinkBlock($this->idLinkBlock);
+
+        $arrayLinkBlock = (array) $linkBlock;
+
+        //The form and the database model don't have the same data hierarchy
+        //Transform array $custom[en][1][name] to $custom[1][en][name]
+        $arrayCustom = [];
+        foreach ($arrayLinkBlock['custom_content'] as $idLang => $customs) {
+            if (empty($customs)) {
+                continue;
+            }
+
+            foreach ($customs as $i => $custom) {
+                $arrayCustom[$i][$idLang] = $custom;
+            }
+        }
+
+        return ['link_block' => [
+            'id_link_block' => $arrayLinkBlock['id_link_block'],
+            'block_name' => $arrayLinkBlock['name'],
+            'id_hook' => $arrayLinkBlock['id_hook'],
+            'cms' => $arrayLinkBlock['content']['cms'],
+            'product' => $arrayLinkBlock['content']['product'],
+            'static' => $arrayLinkBlock['content']['static'],
+            'custom' => $arrayCustom,
+        ]];
+    }
+
+    /**
+     * @param array $data
+     * @return array
+     * @throws \PrestaShop\PrestaShop\Adapter\Entity\PrestaShopDatabaseException
+     */
+    public function setData(array $data)
+    {
+        $linkBlock = $data['link_block'];
+        $errors = $this->validateLinkBlock($linkBlock);
+        if (!empty($errors)) {
+            return $errors;
+        }
+        $customContent = [];
+        if (isset($linkBlock['custom']) && count($linkBlock['custom']) > 0) {
+            foreach ($linkBlock['custom'] as $customIndex => $customLanguages) {
+                foreach ($customLanguages as $idLang => $custom) {
+                    $customContent[$idLang][$customIndex] = $custom;
+                }
+            }
+        }
+
+        if (!isset($linkBlock['id_link_block']) || empty($linkBlock['id_link_block'])) {
+            $linkBlockId = $this->repository->create(
+                $linkBlock['block_name'],
+                $linkBlock['id_hook'],
+                $linkBlock['cms'],
+                $linkBlock['static'],
+                $linkBlock['product'],
+                $customContent
+            );
+        } else {
+            $linkBlockId = $linkBlock['id_link_block'];
+            $this->repository->update(
+                $linkBlockId,
+                $linkBlock['block_name'],
+                $linkBlock['id_hook'],
+                $linkBlock['cms'],
+                $linkBlock['static'],
+                $linkBlock['product'],
+                $customContent
+            );
+        }
+        $this->setIdLinkBlock($linkBlockId);
+        $this->updateHook($linkBlock['id_hook']);
+        $this->cache->clearModuleCache();
+
+        return [];
```

Comment:
You maybe need to return the $linkBlockId, or something to tell everything works fine?

Reply:
this follows the `FormDataProviderInterface` which requires to return an array of errors, empty array means no errors and we already have the if at this point
But I agree this implementation is not adapted, it was meant to be changed (cf the discussion about EntityForm management last week), I wanted to wait until the Core implementation is defined to refactor but since the discussion is blocked and the freeze is for the 20th I am afraid the refacto will have to wait for the 1.7.6
we (sadly) decided this with @mickaelandrieu